### PR TITLE
Remove HirId -> LocalDefId map from HIR.

### PR DESCRIPTION
**Pull request by cjgillot**
Having this map in HIR prevents the creating of new definitions after HIR has been built.
Thankfully, we do not need it.

Based on https://github.com/rust-lang/rust/pull/103902

### CODE REVIEW

File: compiler/rustc_save_analysis/src/lib.rs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #![recursion_limit = "256"]
 #![allow(rustc::potential_query_instability)]
 #![feature(never_type)]
+#![feature(associated_type_defaults)]
```
Comment (WaffleLapkin) [inline compiler/rustc_save_analysis/src/lib.rs:6]: Please do not use associated type defaults, r-a chokes on them, https://github.com/rust-lang/rust-analyzer/issues/13693.

**Review by WaffleLapkin — APPROVED** [review]
I'm not sure what the implications of using different IDs are, but generally looks good to me (especially the perf improvement 👀 ).

r=me after #103902 gets merged, this is rebased on top of master and associated type defaults are removed.

P.S. sorry for so many nits .-.

**Comment by nnethercote** [comment]
The (small) improvements outweigh the (small) regressions.

@rustbot label: +perf-regression-triaged

**Pull request merged**